### ISSUE TITLE
UTF8 support

### ISSUE THREAD

**Issue by victorlevasseur** [4ian/GD#19]
The idea is to create a simple gd::String class providing a string which supports utf8 internally (inside a standard std::string) so that GD would not have to deal explicitly with utf8 problems (such as string size, ...). As Darkhog mentionned, the utf-cpp library (http://utfcpp.sourceforge.net/) can be useful as it provide functions to deal with that encoding.
Utf8 is a good choice because it consumes less memory than Utf32 (used by sf::String) and is directly compatible with the first 128 chars of ASCII.

I don't have informations about how we can handle utf8 in javascript for the web platform, but I think it's managed directly inside.

**Comment by victorlevasseur** [comment]
I'm currently doing some tests on GDevelop for utf8 (on my utf8 branch based on c++11).

It's working fine : JS code works perfectly without changes (just need to add the charset to index.html) and features like text length are working correctly. I'm able to display almost all unicode characters.

On GDCore/GDCpp, I'm able to store UTF8 inside a std::string and I've created some tools for conversions between "UTF8 std::string", sf::String (UTF32) and wxString. So, you need to use these functions everywhere a conversion between wxString and std::string happens because you'll get an ASCII std::string without. Then, functions like string length are invalidated (because of multi-bytes characters) but UTF8-CPP provides functions to get the real size of the string.

**Comment by 4ian** [comment]
You're doing an awesome job as usual! :smile:  I've checked your branch, the file organisation and the naming scheme sounds good and consistent with the rest of the codebase (gd::utf8::xxx is perfect as it clearly enable to find in the code where utf8 is used).
So, so far so good! :+1: 

I think we don't need to implement utf8 support "everywhere": the most important seems to be the Text object :) 
When you'll be done with it, I'll merge it (I have to merge the c++11 branch also before that by the way) and release a test version to make sure that the move to C++11 and utf8 didn't break anything for end users. :)

**Comment by victorlevasseur** [comment]
In fact, it would break string using characters outside the traditional ASCII (I mean from 128 to 255).
I don't agree when you say that only the text object should have UTF8 support : expressions should be UTF8 strings. That way, you can set the text object content with a new UTF8 string.
It's already working on the version available on the utf8 branch (but it needs heavy testing). However, they are some weird things due to the way actions/conditions are declared : I need to show instructions sentences as non UTF8 strings (as they are in the user's locale) but the parameters inside as UTF8 strings (as they may contain special characters). That's why the method ```EventsRenderingHelper::DrawTextInArea``` has a new bool parameter to determine whether the string is encoded in UTF8 or in the user's locale.

**Comment by victorlevasseur** [comment]
To keep compatibility with extensions which rely on "locale" std::string (for example, when loading a file using its filepath), I've added a bool inside gd::ParameterMetadata to say whether the parameter accept an UTF8 string or wants to locale string. If the parameter is not able de receive UTF8 input (default behavior), the result of the string expression is converted from UTF8 to the current locale. (not used in JS)

**Comment by 4ian** [comment]
Well, you surely know better than me what should be done, but I'm a bit afraid that lots of conversion from UTF8 to current locale happens at runtime, which could be bad for performances. What do you think about it?
(Nothing related, but I saw gd::utf8::SubStr and I think it would be nice to add one or two tests cases for it :) It is a perfect candidate for tests)

**Comment by victorlevasseur** [comment]
Maybe there will be less conversion if all instructions use Utf8 by default and we can deactivate it for specials ones ?
or we can adapt the instructions that doesn't support UTF8 to use utf8::To/FromLocaleString.

**Comment by 4ian** [comment]
In which case using UTF8 will break the instructions?

**Comment by victorlevasseur** [comment]
I've only noticed problems with actions using a filepath. For Common dialogs, I've added a conversion function between UTF8 and std::wstring (Windows can use UTF16 wide char)
Other actions seems to be working properly : even variables with things like ```myVar["Unicode string here"]``` works :smile: .

**Comment by 4ian** [comment]
Sounds good! So I think the right approach for instructions is to let everything in UTF8 unless there is a problem (like with file path) in which case we have to add a manual conversion.
This way we avoid lots of potentially costly conversions, and the only changes to be made in the code are in functions dealing with filepath. 

``myVar["Unicode string here"]`` will be awesome! :smile: 
Can I try your UTF8 branch or is there still some work in progress? :)

**Comment by victorlevasseur** [comment]
Good, I've done this changes (Just call CantUseUtf8() after declaring a parameter or DontReturnUtf8 after an string expression).

You can try, but I can't garantuee that an old ASCII project can be opened without problems with this version (I'll see that problem later, maybe converting the whole project into utf8 before opening according to the project file version ?)

**Comment by victorlevasseur** [comment]
So, good news, everything is working thing for CommonDialogs (only tested on Windows). The WinAPI is so cool (irony :smile:).
By the way, we can even use a unicode string as a pure variable name.

**Comment by 4ian** [comment]
Nice! I'm really busy these days so sorry for not being able to be totally at work on it! 
Are you making other changes?

I've tested the branch yesterday and I didn't notice any problem with already existing games :) We'll have to make some tests to ensure that games saved on Windows are correctly opened on Ubuntu (for now, games made on Windows having strings with not ASCII characters (like french diacritics: é...) are interpreted as empty strings).

**Comment by victorlevasseur** [comment]
I still need to adapt some "String manipulation" expressions and make some tests. I'm also creating a group in the doxygen documentation to explain how to use UTF8 strings and where they are used.

**Comment by 4ian** [comment]
Excellent, I'm highly interested in these documentations pages to make clear for everyone how things are working!

**Comment by victorlevasseur** [comment]
I've fixed crash when opening an ANSI project containing locale depend characters (crash inside utf8-cpp because of invalid codepoints when trying to convert to sf::String or wxString).
So, in the last commit, when GD opens a project file with "ISO-8859-1" in its declaration, it converts the whole file in utf8 (using gd::utf8::FromLocaleString) into myProject.gdg.utf8 then opens it (as it was myProject.gdg).

**Comment by victorlevasseur** [comment]
I've fixed the issue. Also, GDevelop force the encoding when loading a project to avoid TinyXML to auto-detect the wrong encoding (some extensions still save ANSI strings and TinyXML might think that the file is an ANSI file).

**Comment by victorlevasseur** [comment]
So, it's a bit weird. Old projects files were already saved in UTF8 in Linux so I've disable the conversion on Linux. But, then, you can't open correctly a Windows ANSI project on GDevelop Linux (as we can't use the "locale string -> UTF8" conversion as the locale is already UTF8) : I've added a message saying that the user should convert the project file to UTF8 using an external text editor before opening it with GDevelop. :(

**Comment by 4ian** [comment]
I guess that it is ok as people opening a project saved in Windows on Ubuntu are quite rare, and as long as they save it again with the new version on Windows it will be ok! So the warning message is good enough :)
Almost all users shouldn't notice the change, but I'll increase the minor version number and warn that projects saved with the next version could be not retro-compatible.

**Comment by victorlevasseur** [comment]
Some news : I'm currently adding a new macro : GD_T which is returns a UTF8 encoded string from a translation. It's defined in GDCore/Tools/Localization.h. Now, _() will not be redefined as it was before : it now returns a wxString and should only be used in GUI.

**Comment by 4ian** [comment]
I'm a bit concerned by this extra complexity: why is it necessary?

**Comment by victorlevasseur** [comment]
Because the trick you were using to get std::string from _() was converting them to the current locale.

**Comment by 4ian** [comment]
Ok... To be clear, the rationale behind the current design was to have a single macro for all translations: this macro was set to nothing when compiling for GDC++ Runtime, meaning that all runtime sentences are in english (that's fine because at runtime almost all translations are for logs and not for the end user). When compiling for the IDE, the macro was set to the standard wx macro to support translation, except that it will be also converted to a std::string to allow its use in non GUI (i.e not wxwidgets related) classes.
Couldn't we convert to std string as before but using wx appropriate functions for handling UTF8?

**Comment by victorlevasseur** [comment]
The problem is that those strings (these are not only the log translation but also all the actions/conditions/expressions strings) are encoded in the current locale. It's different from the other strings (which are in UTF8) and can create errors.

Then, if we redefine _() to get UTF8 string, the conversion to wxString will not work as the string will be considered in the current locale.

**Comment by 4ian** [comment]
So at the end when should GD_T() and _() be used?

Ok, surely something broke in CodeCompiler when adapting it for Mac OS (for now on Mac OS it's broken as linkage expect .dylib files and GD extensions are .xgd files).

**Comment by victorlevasseur** [comment]
GD_T() -> UTF8 std::string (instructions/expressions description, title, object's properties, debugger...)
_() -> wxString (mostly for generated GUI and all wx calls when a wxString is needed)

(you can look on utf8-tr-macro branch to see the changes, not fully working as some std::string are still converted without utf8::ToWxString)

**Comment by 4ian** [comment]
Ok I think you're most advanced on this point than me, I still struggle with these encoding details, so do it as you think it's best :) 
I just want to keep things as simple as possible (that's why I don't like this GD_T ;) ) Note that I the source code, everything is expected to be written using ANSI characters (not utf8 source code) in plain english: don't know if it can be of any help or if it is even related but might be worth keeping it in mind.

**Comment by victorlevasseur** [comment]
It would be fantastic if we could tell wxWidgets to consider strings as UTF8 by default (that way, we could define _() with gd::utf8::FromWxString(wxGetText...(s)) and wxWidgets will do the opposite conversion automatically when needed) but this is not possible... That's why GD_T is, I think, the most appropriate solution. I think it could be more dangerous to keep the localized string in the current locale and the other in UTF8 (and what happen if we want to launch GD in russian on an english Windows ?). Also, there were some complicated stuff when mixing UTF8 and not UTF8 string : it was the case when rendering an instruction's text : parameters were UTF8 string while other part of the instruction sentence was in the current locale, the temporary solution was to add a isUtf8 boolean to know.
But it is up to you, it's your software. ;)

**Comment by 4ian** [comment]
Well, if you think it's the most appropriate solution, let's go for it. ;) 
If I understand correctly, the pitfall with having _() returning a std::string is that wxWidgets assumes that all std::string are in the current locale, that's it?

**Comment by victorlevasseur** [comment]
I may have another solution : _() returning a class which can convert to std::string or wxString automatically.

**Comment by 4ian** [comment]
Or (may sound dumb but who knows) a simple function/macro that:
-does nothing (as now) for GDC++ Runtime.
-returns a std::string that can be converted to a wxString, that is to say, if I understood you correctly, a string that is in current locale. To do so, I imagine that we can construct a wxString using `wxGetTranslation` then convert it to a std::string with the right encoding/locale using a wxString function.

**Comment by victorlevasseur** [comment]
You mean use gd::utf8::FromWxString to get an UTF8 encoded std::string ? 
Yes, but when _() will be used in GUI, wxWidgets will **automatically** convert the std::string to a wxString considering that the string is **in the current locale** (it won't be the case). That' the problem...

**Comment by 4ian** [comment]
I mean do somehow:
`wxGetTranslation() (wxString) --- to locale --> std::string --- from locale ---> wxString`

Is this even feasible?

**Comment by victorlevasseur** [comment]
That's what is currently done, but the problem is that the std::string is in the current locale. That's not what I want (I need actions/conditions/expressions titles, description and many other things in UTF8).

**Comment by 4ian** [comment]
Ok, so in this case continue with what you've done, it's technically correct anyway so let's build something stable and will see later if something can be simplified or not :)

**Comment by victorlevasseur** [comment]
Ok, the macro is working now, I will update some GUI to take care of UTF8 strings (for instructions and expressions description and title, properties' name and description...)

**Comment by 4ian** [comment]
Ok! By the way, is the boolean trick you used for rendering actions/conditions still needed?

**Comment by victorlevasseur** [comment]
No, because all string are now in UTF8.

**Comment by 4ian** [comment]
Excellent! :)

**Comment by 4ian** [comment]
Did you had any time to make some progress on it? Stuck on another problem maybe?
Let me know :)

**Comment by victorlevasseur** [comment]
I dont' have any problems. I just didn't have a lot of time because I had a lot of exams last week and this week too.
Anyway, this feature is a kind of long term feature : there are a lot of things to do and even more to test.

**Comment by 4ian** [comment]
Ok, no problem this is perfectly ok :) I was just afraid that you got discouraged by the task. Good luck with your exam! :smile:

**Comment by victorlevasseur** [comment]
Thanks (but I finished them yesterday ^^). So, now, I'll have more time to develop this feature (and play GTA V ^^).

**Comment by 4ian** [comment]
Cool cool! :D (Ah GTA V, I still have to get it on my PC, the game looks awesome!)

**Comment by victorlevasseur** [comment]
I think we can close this issue now as UTF8 support is now stable on v4 branch.
